In [12]:
#1 a)


data = [
{"employeeID": 0, "name": "Alice", "department": "HR", "salary": 31000},
{"employeeID": 1, "name": "Bob", "department": "IT", "salary": 41000},
{"employeeID": 2, "name": "Charlie", "department": "HR", "salary": 28000}]


In [13]:
#b)

data[1]

{'employeeID': 1, 'name': 'Bob', 'department': 'IT', 'salary': 41000}

In [14]:
#c)

for employee in data:
    result = f"{employee['name']} works in department {employee['department']} and earns {employee['salary']}."
    print(result)


Alice works in department HR and earns 31000.
Bob works in department IT and earns 41000.
Charlie works in department HR and earns 28000.


In [15]:
#d) Find out the avarage salary in the company

# Ta ut alla löner från dictionaries och lägger dem i en lista (salarries)
salarries = [employee['salary'] for employee in data]

# Beräkna medelvärdet av lönerna
avarage_salary = sum(salarries) / len(salarries)

print(f"medelvärdet av lönerna är {avarage_salary}")

medelvärdet av lönerna är 33333.333333333336


In [16]:
#e) avarage salary in the hr department

selected_dep = "HR"

# Ta ut lönerna för de anställda i den valda avdelningen
salaries = [employee['salary'] for employee in data if employee['department'] == selected_dep]

hr_avarage = sum(salaries) / len(salaries)

hr_avarage


29500.0

In [17]:
#f) extract all records where the department is HR

# skapar en ny lista hr_employees som bara innehåller dictionaries där department är "HR".
hr_employees = [employee for employee in data if employee['department'] == 'HR']
hr_employees


[{'employeeID': 0, 'name': 'Alice', 'department': 'HR', 'salary': 31000},
 {'employeeID': 2, 'name': 'Charlie', 'department': 'HR', 'salary': 28000}]

In [18]:
#g)

# employee['department'] for employee in data} skapar en mängd (set) med alla unika department-värden från listan data. Mängder lagrar bara unika värden, så dubbletter filtreras automatiskt bort.

unique_dep = {employee['department'] for employee in data}
unique_dep

{'HR', 'IT'}

In [19]:
#h)

# Lägg till en ny kolumn 'age' med specifika värden

# enumerate(data): Denna funktion returnerar både index och dictionary för varje element i listan data.
# If-sats: Beroende på indexet sätts värdet för nyckeln "age" till 26, 37 eller None

for index, employee in enumerate(data):
    if index == 0:
        employee['age'] = 26
    elif index == 1:
        employee['age'] = 37
    else:
        employee['age'] = None

for employee in data:
    print(employee)

{'employeeID': 0, 'name': 'Alice', 'department': 'HR', 'salary': 31000, 'age': 26}
{'employeeID': 1, 'name': 'Bob', 'department': 'IT', 'salary': 41000, 'age': 37}
{'employeeID': 2, 'name': 'Charlie', 'department': 'HR', 'salary': 28000, 'age': None}


In [20]:
#i)

def add_personel(employeeID, name, department, salary, age=None):
     # Skapa en ny dictionary för den nya medarbetaren
     new_employee = {
        "employeeID": employeeID,
        "name": name,
        "department": department,
        "salary": salary,
        "age": age
    }
     # Lägg till den nya medarbetaren i listan
     data.append(new_employee)

# Lägg till en ny medarbetare
add_personel(3,"David", "IT", 35000, 30)

for employee in data:
     print(employee)


{'employeeID': 0, 'name': 'Alice', 'department': 'HR', 'salary': 31000, 'age': 26}
{'employeeID': 1, 'name': 'Bob', 'department': 'IT', 'salary': 41000, 'age': 37}
{'employeeID': 2, 'name': 'Charlie', 'department': 'HR', 'salary': 28000, 'age': None}
{'employeeID': 3, 'name': 'David', 'department': 'IT', 'salary': 35000, 'age': 30}


In [21]:
#1 Define a Pokemon class with attributes for name, type, and level. Instantiate a pokemon named Pikachu and give it an electric type with level 10.

class Pokemon:
    def __init__(self, name: str, type_: str, level: int) -> None:
        self.name = name
        self.type = type_
        self.level = level

p1 = Pokemon("Pikachu", "electric", 10)

print(p1.name, p1.type, p1.level)


Pikachu electric 10


In [25]:
#b Implement `__str__()` method to display the Pokemon's details. Also implement a `__repr__()` method.

#f)  f) To implement battling we need to do the following: 
# - add an attribute hp (healthpoints) to Pokemon class.
# - implement a (pseudo)private method take_damage that subtracts the damage on the healthpoints
# - implement an attack method in the Pokemon class that takes in another Pokemon as an argument. It inflicts damage on other pokemon based on its own level times 2. Print out damage taken and life points left on the pokemon. When damage exceeds the life points, the pokemon is fainted.

# g) Instantiate two pokemons and let them battle until one has fainted and announce a winner. They take turns in fighting one attack at a time.


class Pokemon:
    def __init__(self, name: str, type_: str, level: int, hp: int) -> None:
        self.name = name
        self.type = type_ # använder type_ för att inte orsaka problem med pythons inbyggda metod
        self.level = level
        self.hp = hp
    
    # Returnerar en användarvänlig strängrepresentation som är lätt att läsa. Används t.ex. när du skriver ut objektet med print.
    def __str__(self):
        return f"{self.name} {self.type} {self.level} {self.hp}"

    # Returnerar en mer detaljerad och formell strängrepresentation som kan användas för att skapa objektet igen. Används främst för felsökning och vid utveckling.
    def __repr__(self):
        return f"Pokemon(name='{self.name}', type_='{self.type}', level={self.level}, hp={self.hp})"

    def __take_damage(self, damage: int) -> None:
        if damage > 0:
            self.hp -= damage
            if self.hp < 0:
                self.hp = 0
                
    
    def attack(self, other_pokemon: Pokemon) -> None:
        # Beräkna skada baserat på Pokémon nivå
        damage = self.level * 2
        
        # Skriv ut detaljer om attacken
        print(f"{self.name} attacks {other_pokemon.name} with {damage} damage!")

        # skada på den attackerade Pokémon
        other_pokemon.__take_damage(damage)

        # Skriv ut skadan och återstående HP
        if other_pokemon.hp == 0:
            print(f"{other_pokemon.name} has fainted!")
        else:
            print(f"{other_pokemon.name} took {damage} damage and has {other_pokemon.hp} HP left.")

    def battle(pokemon1: Pokemon, pokemon2: Pokemon) -> None:
    
        turn = 0  # Bestäm vilken Pokémon som ska attackera
        
        while pokemon1.hp > 0 and pokemon2.hp > 0:
            if turn % 2 == 0:
                # Pokémon 1 attackerar Pokémon 2
                pokemon1.attack(pokemon2)
            else:
                # Pokémon 2 attackerar Pokémon 1
                pokemon2.attack(pokemon1)
            
            turn += 1  # Byt tur
        
        # Tillkännage vinnaren
        if pokemon1.hp > 0:
            print(f"{pokemon1.name} vinner!")
        else:
            print(f"{pokemon2.name} vinner!")     


p1 = Pokemon("Pikachu", "electric", 10, 100)
p2 = Pokemon("squritle", "water", 6, 100)

p1.attack(p2)
Pokemon.battle(p2, p1)

Pikachu attacks squritle with 20 damage!
squritle took 20 damage and has 80 HP left.
squritle attacks Pikachu with 12 damage!
Pikachu took 12 damage and has 88 HP left.
Pikachu attacks squritle with 20 damage!
squritle took 20 damage and has 60 HP left.
squritle attacks Pikachu with 12 damage!
Pikachu took 12 damage and has 76 HP left.
Pikachu attacks squritle with 20 damage!
squritle took 20 damage and has 40 HP left.
squritle attacks Pikachu with 12 damage!
Pikachu took 12 damage and has 64 HP left.
Pikachu attacks squritle with 20 damage!
squritle took 20 damage and has 20 HP left.
squritle attacks Pikachu with 12 damage!
Pikachu took 12 damage and has 52 HP left.
Pikachu attacks squritle with 20 damage!
squritle has fainted!
Pikachu vinner!


: 

In [ ]:
#c ) Create a subclass WaterPokemon that inherits from Pokemon and add a unique method splash. This method prints out the pokemons name and that it splashes some water.


# Definiera en subklass WaterPokemon som ärver från Pokemon
class WaterPokemon(Pokemon):        
    # Anropa bas-klassens __init__ metod för att initialisera arvda attribut    
    def __init__(self, name: str, type_: str, level: int, hp: int) -> None:
        super().__init__(name, type_, level, hp)

    def splash(self) -> None:
        print(f"{self.name} splashes some water")

p2 = WaterPokemon("Squirtle", "water", 5, 20)

print(p2)
p2.splash()



Squirtle water 5 20
Squirtle splashes some water


In [ ]:
#d) d) Now instantiate two Squirtles and a Polywag. Squirtles and Polywags are of type WaterPokemon.

Squirtle1 = WaterPokemon("Squirtle1", "water", 10)
Squirtle2 = WaterPokemon("Squirtle2", "water", 8)
Polywag = WaterPokemon("Polywag1", "water", 15)

print(Squirtle1)
print(Squirtle2)
print(Polywag)
Polywag.splash()

Squirtle1 water 10
Squirtle2 water 8
Polywag1 water 15
Polywag1 splashes some water


In [ ]:
#e) e) Loop through a tuple of all your instantiated pokemons and print them out.

pokes = p1, Squirtle1, Squirtle2, Polywag
pokes

for pokemons in pokes:
    print(pokemons)

Pikachu electric 10
Squirtle1 water 10
Squirtle2 water 8
Polywag1 water 15


### 2 Analyze a log file

It is common to log from different softwares and systems so that you always can go back to find out what happened, e.g. for debugging purposes. Under the data directory, there is a short log from a data engineering tool called data build tool (dbt), which is used for data transformations with SQL and jinja templating language. 

There seem to have been some errors during runs. You are tasked to read this log using Python and parse it to find some useful information for debugging. Print them out in a nicely formatted way. 

&nbsp; a) Find specific error messages and codes to understand the nature of the issues encountered.

&nbsp; b) Identify timestamps when the errors happened. 

&nbsp; c) Find performance metrics.

&nbsp; d) Find application information.

&nbsp; e) Are there some other valuable information that can be extracted the log?


In [12]:
from parse import parse

with open("../Data/dbt.log", "r") as handle:
    data = handle.readlines()

data[:15]

['\n',
 '\x1b09:33:53.284167 [debug] [ThreadPool]: Opening a new connection, currently in state closed\n',
 '\x1b09:33:53.934267 [debug] [ThreadPool]: Snowflake adapter: Snowflake error: 003001 (42501): SQL access control error:\n',
 "Insufficient privileges to operate on database 'JOB_ADS'\n",
 '\x1b09:33:53.934836 [debug] [ThreadPool]: Snowflake adapter: Error running SQL: macro create_schema\n',
 '\x1b09:33:53.935193 [debug] [ThreadPool]: Snowflake adapter: Rolling back transaction.\n',
 '\x1b09:33:53.935611 [debug] [ThreadPool]: On create_job_ads_warehouse_warehouse: Close\n',
 '\x1b09:33:54.072516 [debug] [ThreadPool]: Re-using an available connection from the pool (formerly create_job_ads_warehouse_warehouse, now create_job_ads_warehouse_marts)\n',
 '\x1b09:33:54.073564 [debug] [ThreadPool]: Creating schema "database: "job_ads"\n',
 'schema: "warehouse_marts"\n',
 '"\n',
 '\x1b09:33:54.077179 [debug] [ThreadPool]: Using snowflake connection "create_job_ads_warehouse_marts"\n',
 '

Förklaring av nedan kod:

1. Import av biblioteket re:

    re används för att hantera reguljära uttryck, vilket gör det möjligt att söka efter specifika mönster i texten, till exempel för att extrahera felkoder eller tidsstämplar från loggfilen.

2. Funktion parsa_loggfil(fil_sökväg):

    Denna funktion tar emot sökvägen till en loggfil som argument och går igenom filen rad för rad för att extrahera och lagra olika typer av information.

3. Initiering av listor och dictionaries:

    felmeddelanden: En lista för att lagra alla felmeddelanden och deras associerade felkoder.
    tidsstämplar: En lista för att lagra tidsstämplar från loggfilen.
    prestandamått: En ordbok för att lagra olika prestandamått som exekveringstid och minnesanvändning.
    app_info: En ordbok för att lagra information om applikationen, som exempelvis versionen av dbt.
    annan_info: En lista för att lagra annan viktig information som inte faller under de övriga kategorierna, som anslutningsstatus.

4. Läsa loggfilen rad för rad:

    Filen öppnas och läses rad för rad. Varje rad analyseras för att identifiera relevant information.

5. Uppgift a: Extrahera felmeddelanden och koder:

    Om en rad innehåller ordet "error", försöker koden hitta en matchande felkod (t.ex. från Snowflake) och lagrar denna tillsammans med tillhörande meddelande och tidsstämpel i felmeddelanden.

6. Uppgift b: Extrahera tidsstämplar:

    Om en rad börjar med ett mönster som liknar en tidsstämpel (t.ex. "09:33:53.284167"), lagras denna i listan tidsstämplar.

7. Uppgift c: Extrahera prestandamått:

    Om en rad innehåller "Resource report", extraheras olika prestandamått som exekveringstid och minnesanvändning, vilka lagras i ordboken prestandamått.

8. Uppgift d: Extrahera applikationsinformation:

    Om en rad innehåller information om applikationen, t.ex. versionen av dbt, extraheras denna och lagras i app_info.

9. Uppgift e: Extrahera annan värdefull information:

    Information som rör öppnande och stängande av anslutningar lagras i listan annan_info.

10. Returnera extraherade data:

    Alla extraherade data returneras som en ordbok, vilket gör det lätt att använda resultaten senare i koden.

11. Användning av parsa_loggfil:

    Funktionen parsa_loggfil anropas med sökvägen till loggfilen, och de extraherade data lagras i variabeln logg_data.

12. Utskrift av resultaten:

    Resultaten från de olika uppgifterna skrivs ut i en organiserad och läsbar form:
    Felmeddelanden: Visar tidsstämpel, felkod (om någon), och meddelande.
    Tidsstämplar: Visar alla extraherade tidsstämplar.
    Prestandamått: Visar olika prestandamått som exekveringstid och använd minnesmängd.
    Applikationsinformation: Visar detaljer som version av dbt och registrerade adapters.
    Annan information: Visar annan viktig information, såsom status på anslutningar.

In [17]:
import re

# Funktion för att parsa loggfilen
def parsa_loggfil(fil_sökväg):
    # Skapa tomma listor och dicts för att lagra resultat
    felmeddelanden = []  # För att lagra felmeddelanden och felkoder
    tidsstämplar = []  # För att lagra tidsstämplar
    prestandamått = {}  # För att lagra prestandamått (som exekveringstid)
    app_info = {}  # För att lagra information om applikationen (som version)
    annan_info = []  # För att lagra annan värdefull information

    # Öppna och läs loggfilen rad för rad
    with open(fil_sökväg, "r") as fil:
        for rad in fil:
            # Uppgift a: Extrahera felmeddelanden och koder
            if "error" in rad.lower():
                felkod_match = re.search(r'Snowflake error: (\d+)', rad)
                if felkod_match:
                    felkod = felkod_match.group(1)
                    felmeddelanden.append({"tidsstämpel": rad[:15], "felkod": felkod, "meddelande": rad.strip()})
                else:
                    felmeddelanden.append({"tidsstämpel": rad[:15], "meddelande": rad.strip()})

            # Uppgift b: Extrahera tidsstämplar
            tidsstämpel_match = re.match(r'^\d{2}:\d{2}:\d{2}\.\d{6}', rad)
            if tidsstämpel_match:
                tidsstämplar.append(tidsstämpel_match.group(0))

            # Uppgift c: Extrahera prestandamått
            if "Resource report" in rad:
                prestandamått["command_wall_clock_time"] = re.search(r'"command_wall_clock_time": ([\d\.]+)', rad).group(1)
                prestandamått["process_user_time"] = re.search(r'"process_user_time": ([\d\.]+)', rad).group(1)
                prestandamått["process_kernel_time"] = re.search(r'"process_kernel_time": ([\d\.]+)', rad).group(1)
                prestandamått["process_mem_max_rss"] = re.search(r'"process_mem_max_rss": "([\d]+)', rad).group(1)

            # Uppgift d: Extrahera applikationsinformation
            if "Running with dbt" in rad:
                app_info["dbt_version"] = re.search(r'dbt=(\d+\.\d+\.\d+)', rad).group(1)

            # Uppgift e: Extrahera annan värdefull information
            if "Registered adapter" in rad:
                app_info["adapter"] = re.search(r'Registered adapter: ([a-zA-Z0-9_]+)', rad).group(1)

            # Extrahera information om öppnande/stängande av anslutningar
            if "Opening a new connection" in rad or "Closing connection" in rad:
                annan_info.append(rad.strip())

    # Returnera de extraherade data som en ordbok
    return {
        "felmeddelanden": felmeddelanden,
        "tidsstämplar": tidsstämplar,
        "prestandamått": prestandamått,
        "app_info": app_info,
        "annan_info": annan_info
    }

# Exempel på användning av funktionen
logg_data = parsa_loggfil("../Data/dbt.log")

# Uppgift a: Skriv ut felmeddelanden
print("Felmeddelanden:")
for fel in logg_data["felmeddelanden"]:
    print(f"- Tidsstämpel: {fel['tidsstämpel']}, Felkod: {fel.get('felkod', 'Ingen kod')}, Meddelande: {fel['meddelande']}")

# Uppgift b: Skriv ut tidsstämplar
print("\nTidsstämplar för fel:")
for ts in logg_data["tidsstämplar"]:
    print(f"- {ts}")

# Uppgift c: Skriv ut prestandamått
print("\nPrestandamått:")
for nyckel, värde in logg_data["prestandamått"].items():
    print(f"- {nyckel}: {värde}")

# Uppgift d: Skriv ut applikationsinformation
print("\nApplikationsinformation:")
for nyckel, värde in logg_data["app_info"].items():
    print(f"- {nyckel}: {värde}")

# Uppgift e: Skriv ut annan information
print("\nAnnan information:")
for info in logg_data["annan_info"]:
    print(f"- {info}")

Felmeddelanden:
- Tidsstämpel: 09:33:53.93, Felkod: 003001, Meddelande: 09:33:53.934267 [debug] [ThreadPool]: Snowflake adapter: Snowflake error: 003001 (42501): SQL access control error:
- Tidsstämpel: 09:33:53.93, Felkod: Ingen kod, Meddelande: 09:33:53.934836 [debug] [ThreadPool]: Snowflake adapter: Error running SQL: macro create_schema
- Tidsstämpel: 09:33:54.75, Felkod: 003001, Meddelande: 09:33:54.755169 [debug] [ThreadPool]: Snowflake adapter: Snowflake error: 003001 (42501): SQL access control error:
- Tidsstämpel: 09:33:54.75, Felkod: Ingen kod, Meddelande: 09:33:54.756174 [debug] [ThreadPool]: Snowflake adapter: Error running SQL: macro create_schema
- Tidsstämpel: 09:33:54.90, Felkod: Ingen kod, Meddelande: 09:33:54.908251 [error] [MainThread]: Encountered an error:
- Tidsstämpel: Database Error
, Felkod: Ingen kod, Meddelande: Database Error
- Tidsstämpel:   003001 (42501, Felkod: Ingen kod, Meddelande: 003001 (42501): SQL access control error:

Tidsstämplar för fel:

Pres

# 3. Theory questions

These study questions are good to get an overview of how snowflake works.

a) What is the difference between dictionary and lists?

b) What are postional arguments in functions?

c) What is the difference between using postional arguments and keyword arguments in functions?

d) What is the difference between a list and a tuple?

e) Explain the concept of Python's dynamic typing.

f) What is the purpose behind the `__repr__()` method in a class?

g) How does `__str__()` method differ from `__repr__()` method?

i) What are the pillars of OOP?

# 3

a) En lista är som en rad med saker som är ordnade i en viss ordning. man kan plocka ut en sak från listan med hjälp av dess plats (index).
   En dict är som en samling lådor där varje låda har en etikett (nyckel, key) och varje etikett pekar på en viss sak (värde, value) i lådan.

    Lista: Saker är i en viss ordning, och du hämtar dem med deras plats.
    Dictionary: Saker är inte i ordning, och du hämtar dem med deras etikett (nyckel).

b) Positional arguments i funktioner är argument som skickas till en funktion baserat på deras position i funktionsanropet. När du kallar en funktion med positional arguments, måste du ange värdena i samma ordning som   funktionens parametrar är definierade.    

c) Skillnaden är att i positional så använder man deras position tex (person("anna", 25)) medans i keyword så skriver man in deras nyckelord före tex (person(ålder=25, namn="anna"))

d) En lista är muterbar det är inte en tuple. lista använder [] medans tuple använder (). tuple är snabbare och mer minneseffektiv

e) Python använder sig av dynamisk typning, vilket innebär att du inte behöver ange vilken datatyp en variabel har när du skapar den. Istället bestäms datatypen automatiskt baserat på det värde du tilldelar variabeln.
Eftersom Python är dynamiskt typad kan du ändra typen av en variabel genom att tilldela ett nytt värde av en annan typ.

f) __repr__()-metoden i en klass används för att ge en strängrepresentation av ett objekt, speciellt för utvecklare. Det är som en beskrivning av objektet som är tänkt att vara tydlig och användbar när man felsöker eller vill se vad objektet innehåller.

g) __str__(): Används för att ge en lättläst och användarvänlig beskrivning av ett objekt. Den är tänkt för att visa en mer "mänsklig" representation, exempelvis vid utskrift till skärmen.
Kortfattat: __str__() är för användare, medan __repr__() är för utvecklare.

i) De fyra grundpelarna i objektorienterad programmering (OOP) är:

Inkapsling (Encapsulation):
Inkapsling innebär att dölja en objekts inre tillstånd och endast exponera vissa delar via publika metoder. Detta skyddar objektets data och säkerställer att det används på rätt sätt.

Arv (Inheritance):
Arv gör det möjligt att skapa nya klasser baserat på redan existerande klasser. Den nya klassen, som kallas subklass eller barnklass, ärver egenskaper och beteenden från den befintliga klassen, som kallas superklass eller förälderklass.

Polymorfism (Polymorphism):
Polymorfism innebär att samma metod eller funktion kan bete sig olika beroende på vilket objekt det används på. Det tillåter att använda en enhetlig metodsignatur för olika datatyper eller klasser.

Abstraktion (Abstraction):
Abstraktion innebär att fokusera på vad ett objekt gör, snarare än hur det gör det. Genom abstraktion kan man skapa förenklade modeller av komplexa system, där bara relevanta detaljer visas och irrelevanta döljs.








## Glossary

Fill in this table either by copying this into your own markdown file or copy it into a spreadsheet if you feel that is easier to work with.

| terminology              | explanation |
| ------------------------ | ----------- |
| dictionary               |             |
| list                     |             |
| record                   |             |
| list comprehension       |             |
| append                   |             |
| field                    |             |
| row                      |             |
| column                   |             |
| dictionary comprehension |             |
| statement                |             |
| argument                 |             |
| boolean                  |             |
| class                    |             |
| conditional              |             |
| exception                |             |
| function                 |             |
| keyword arguments        |             |
| module                   |             |
| object                   |             |
| return statement         |             |
| tuple                    |             |
| type                     |             |
| self                     |             |
| instatiate               |             |
| iterating                |             |
| encapsulation            |             |
|                          |             |
|                          |             |
|                          |             |